lib/Dialect/Arc/Transforms/LegalizeStateUpdate.cpp

```diff
@@ -126,128 +87,153 @@ struct AccessAnalysis : public DataFlowAnalysis {
-  }
 
-  // Remove any information about locally-defined state which we cannot access
-  // outside the current block. This prevents significant blow-up of the access
-  // sets, since local state accesses don't get transported to parent ops where
-  // they have no meaning.
-  for (auto state : localState) {
-    innerAccesses.remove(state, AccessState::Read);
-    innerAccesses.remove(state, AccessState::Write);
+    // Track the relationship between callers and callees.
+    if (auto callOp = dyn_cast<CallOpInterface>(opNode.op))
+      if (auto calleeOp = callOp.resolveCallable(&symbolTable))
+        get(calleeOp).callers.insert(&opNode);
+
+    // Create the seed accesses.
+    if (auto readOp = dyn_cast<StateReadOp>(opNode.op))
+      add(opNode, Access(readOp.getState(), AccessType::Read));
+    else if (auto writeOp = dyn_cast<StateWriteOp>(opNode.op))
+      add(opNode, Access(writeOp.getState(), AccessType::Write));
   }
+  LLVM_DEBUG(llvm::dbgs() << "- Prepared " << blockAccesses.size()
+                          << " block and " << opAccesses.size()
+                          << " op lattice nodes\n");
+  LLVM_DEBUG(llvm::dbgs() << "- Worklist has " << opWorklist.size()
+                          << " initial ops\n");
+
+  // Propagate accesses through calls.
+  while (!opWorklist.empty()) {
+    auto &opNode = *opWorklist.pop_back_val();
+    if (opNode.callers.empty())
+      continue;
+    auto calleeOp = dyn_cast<CallableOpInterface>(opNode.op);
+    if (!calleeOp)
+      continue;
+    LLVM_DEBUG(llvm::dbgs() << "- Updating callable " << opNode.op->getName()
+                            << " " << opNode.op->getAttr("sym_name") << "\n");
 
-  // Track block argument accesses in a separate analysis state.
-  auto *argAccesses = getOrCreate<ArgumentAccessState>(block);
-  auto result = ChangeResult::NoChange;
-  for (auto arg : block->getArguments()) {
-    if (innerAccesses.remove(arg, AccessState::Read) == ChangeResult::Change)
-      result |= argAccesses->add(arg, AccessState::Read);
-    if (innerAccesses.remove(arg, AccessState::Write) == ChangeResult::Change)
-      result |= argAccesses->add(arg, AccessState::Write);
+    auto &calleeRegion = *calleeOp.getCallableRegion();
+    auto *blockNode = lookup(&calleeRegion.front());
+    if (!blockNode)
+      continue;
+    auto calleeArgs = blockNode->block->getArguments();
+
+    for (auto *callOpNode : opNode.callers) {
+      LLVM_DEBUG(llvm::dbgs() << "  - Updating " << *callOpNode->op << "\n");
+      auto callArgs = cast<CallOpInterface>(callOpNode->op).getArgOperands();
+      for (auto [calleeArg, callArg] : llvm::zip(calleeArgs, callArgs)) {
+        if (blockNode->argAccesses.contains({calleeArg, AccessType::Read}))
+          add(*callOpNode, {callArg, AccessType::Read});
+        if (blockNode->argAccesses.contains({calleeArg, AccessType::Write}))
+          add(*callOpNode, {callArg, AccessType::Write});
+      }
+    }
```

Comment:
Is it correct that this assumes no unstructured control flow (i.e., multiple blocks per region)? If so, should we emit an error of multiple blocks are present?

Reply:
Yeah good point. The pass currently doesn't handle state pointers flowing among ops and blocks. It only supports trivial pass-down of states. I can make it error if a state pointer isn't trivially a block arg or an alloc op.